refactor: update weekly schedule handling for Sonoff TRVZB to use individual day properties

abaddon opened the pull request:
Update weekly schedule handling for Sonoff TRVZB to use individual day properties

From:
``` 
"weekly_schedule": {
        "friday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
        "monday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
        "saturday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
        "sunday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
        "thursday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
        "tuesday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
        "wednesday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17"
},
```

To:
```
"weekly_schedule_monday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
"weekly_schedule_tuesday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
"weekly_schedule_wednesday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
"weekly_schedule_thursday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
"weekly_schedule_friday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
"weekly_schedule_saturday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
"weekly_schedule_sunday": "00:00/16 06:30/19 19:00/20 21:30/17 21:30/17 21:30/17",
```

acommuni commented:
Why did you do this? It was working fine before. Plus, the data was well organized. A flat structure is much less readable.

carbontrader0 commented:
Setting multiple days in one payload does not seem to work for me using this new method. It will set the first attribute (e.g. monday in the example below, but not Tuesday). Seems to need to be done one at a time??

e.g.:
  - action: mqtt.publish
    data:
      topic: zigbee2mqtt/AtticBathroomTRV/set
      payload: >-
        { "weekly_schedule_monday": "00:00/16 08:00/17.5 11:15/16 17:00/17.5 22:30/17 23:15/16",
          "weekly_schedule_tuesday": "00:00/16 08:00/17.5 11:15/16 17:00/17.5 22:30/17 23:15/16" }

Same result when using a single line payload in the Zigbee2MQTT tools dev console. Anything after the comma is ignored
Any help would be appreciated
Thanks
Andy

jamesonuk commented:
Fine that it has changed but what has happened to existing data?
From what I can see in the PR it looks like it is still the same underlying Zigbee structure just split out into separate days rather than one large composite.

Issue is that where weekly schedules were set previously the entries are all blank now so in state I see
```json
    "weekly_schedule": {
        "friday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15",
        "monday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15",
        "saturday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15",
        "sunday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15",
        "thursday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15",
        "tuesday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15",
        "wednesday": "00:00/12 14:00/15 14:00/15 14:00/15 14:00/15 14:00/15"
    },
    "weekly_schedule_friday": null,
    "weekly_schedule_monday": null,
    "weekly_schedule_saturday": null,
    "weekly_schedule_sunday": null,
    "weekly_schedule_thursday": null,
    "weekly_schedule_tuesday": null,
    "weekly_schedule_wednesday": null
```
    
but the data must be stored on the device itself so presumably is actually still set to the old schedule even though it isn't showing?

abaddon commented:
> > If you have a better approach please let me know.
> 
> 
> 
> Splitting into single schedules for each single day makes sense from Home Assistant point of view. 
> 
> 
> 
> What's probably missing is the (removed) option to write the full schedule in one message. This could be implemented
> 
> using the new namings as 2 dimensional Json and loop thru the elements. This would give the option to send a single day, but also to provide all of them.
> 
> 

Yes, I believe it's something I can add. I'll work on it next week. Thx for the suggestion.
> 
> Additionally:
> 
> - Remove the old element "weekly schedule" as this is not updated anymore to have consistency
> 

It doesn’t exist.  If you reset the device state or remove and re-add it from Z2MQTT the new JSON structure no longer contains a weekly_schedule.
I do't know if exist a way to force somehow the reset of the existing json structure or force the delete of the weekly_schedule element.
> - fill the 7 elements for the single weekdays. Currently it's null until a new value was sent thru MQTT
> 
> 
It is populated when Z2MQTT requesting to the device the current state of the device, like a new initialization.

It happens only for the old devices, but not for the new ones.
> 
> weekly_schedule is not updated anymore and keeps old values. This is confusing and would be redundant
> 
> ```
> 
>     },
> 
>     "valve_closing_degree": 100,
> 
>     "valve_closing_limit_voltage": 2842,
> 
>     "valve_motor_running_voltage": 1405,
> 
>     "valve_opening_degree": 20,
> 
>     "valve_opening_limit_voltage": 1822,
> 
>     "weekly_schedule": {
> 
>         "friday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12",
> 
>         "monday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12",
> 
>         "saturday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12",
> 
>         "sunday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12",
> 
>         "thursday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12",
> 
>         "tuesday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12",
> 
>         "wednesday": "00:00/12 00:00/12 00:00/12 00:00/12 00:00/12 00:00/12"
> 
>     },
> 
>     "weekly_schedule_friday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>     "weekly_schedule_monday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>     "weekly_schedule_saturday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>     "weekly_schedule_sunday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>     "weekly_schedule_thursday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>     "weekly_schedule_tuesday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>     "weekly_schedule_wednesday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11"
> 
> }
> 
> ```
> 
> 
> 
> Initial values are set to null. It should be filled at startup. This is also reflected in Home Assistant wrongly.

As explained before, this behavior happens only for the old devices.

I don't know if exist a way to force zigbee2mqtt to re-initialize the existing devices.

For my home I deleted each valve and added it again to resolve it.
> 
> ```
> 
>     "weekly_schedule_friday": null,
> 
>     "weekly_schedule_monday": null,
> 
>     "weekly_schedule_saturday": null,
> 
>     "weekly_schedule_sunday": null,
> 
>     "weekly_schedule_thursday": null,
> 
>     "weekly_schedule_tuesday": null,
> 
>     "weekly_schedule_wednesday": null
> 
> ```
> 
> 
> 
> MQTT message for single day
> 
> ```
> 
> {
> 
>     "weekly_schedule": {
> 
>         "weekly_schedule_sunday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11"
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> MQTT message for all. 
> 
> ```
> 
> {
> 
>     "weekly_schedule": {
> 
>         "weekly_schedule_monday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>         "weekly_schedule_tuesday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>         "weekly_schedule_wednesday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>         "weekly_schedule_thursday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>         "weekly_schedule_friday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>         "weekly_schedule_saturday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11",
> 
>         "weekly_schedule_sunday": "00:00/11 00:00/11 00:00/11 00:00/11 00:00/11 00:00/11"
> 
>     }
> 
> }
> 
> ```
